### DISCUSSION TITLE
Adding vtkImplicitPlaneWidget to scene

### DISCUSSION THREAD

**Discussion by loisvazquezind** [Kitware/trame#156] [Q&A]
Hello.

I'm trying to integrate the vtkImplicitPlaneWidget with the intention of making cuts in the mesh in a visual way for the final user, with a switch button to show/hide it. Also, I have a slider field that is related to the plane, so that it moves according to the slider. In the following image, I show the slider in two different positions and the plane that is drawn based on that slider (note that I am also showing the mesh with low opacity, simply in order to position the plane more intuitively, but it does not affect the current problem).

![slice-plane-img](https://user-images.githubusercontent.com/101557949/203551523-f4dbadd5-93d8-40ce-84ec-4f25e5944d10.jpg)

In short, the workflow would be:
1. Activate the plane with the switch button
2. Move the plane through the slider field to the position where I want to make the cut
3. Click on APPLY to apply a 2D cut where the plane was positioned
4. Deactivate the plane with the switch button

I have two problems:
1. **I can't get step 4 to work**. That is, I get the plane to show up with the switch on (step 1) but **I can't get it to hide when the switch is off** (step 4)... The method related to the switch is:
  ```
  @state.change("show_slice_plane")
  def update_show_slice_plane(show_slice_plane, **kwargs):
      logging.debug(f"Updating show slice plane to '{show_slice_plane}'")
      if show_slice_plane:
          planeWidget.On()
          slice_plane_actor.VisibilityOn()
      else:
          planeWidget.Off()
          slice_plane_actor.VisibilityOff()
          slice_plane_actor.SetVisibility(False)
      ctrl.view_update()
  ```
  
  The code that defines the plane widget is:
  ```
  # Plane Widget
  plane_mapper = vtkDataSetMapper()
  plane_mapper.SetInputConnection(reader.GetOutputPort())	# reader = vtkXMLUnstructuredGridReader()
  
  slice_plane_actor = vtkLODActor()
  slice_plane_actor.SetMapper(plane_mapper)
  slice_plane_actor.GetProperty().SetColor(0, 1, 0)
  slice_plane_actor.SetScale(1.01, 1.01, 1.01)
  renderer.AddActor(slice_plane_actor)
  
  planeWidget = vtkImplicitPlaneWidget()
  planeWidget.SetInteractor(renderWindowInteractor)	# renderWindowInteractor = vtkRenderWindowInteractor()
  planeWidget.SetPlaceFactor(1.25)
  planeWidget.GetPlaneProperty().SetOpacity(0.4)
  planeWidget.SetInputConnection(reader.GetOutputPort())	# reader = vtkXMLUnstructuredGridReader()
  planeWidget.SetOrigin(slice_plane.GetOrigin())	# slice_plane = vtkPlane()
  planeWidget.SetNormal(slice_plane.GetNormal())
  planeWidget.PlaceWidget()
  ```
  What I do get with this code is that the plane does not move with the slider, that is, its functionality is somehow disabled but it is not visually hidden:
  
![slice-plane-disabled](https://user-images.githubusercontent.com/101557949/203551832-7ee40012-c5ef-4e0d-bc06-7fbf8311c679.png)

2. I would like to be able to **configure the plane to be able to move it interactively** using the arrow shown with the plane itself, so that I can do step 2 interactively with the plane instead of using a slider. To do this, I have tried to replicate the code, without success, that I found at https://kitware.github.io/vtk-examples/site/Cxx/Widgets/ImplicitPlaneWidget2/. Then I found [this code](https://www.weiy.city/2019/04/vtk-vtkclippolydata-example/), but I can't integrate it into my workflow either. The problem I'm having is that I can never move the plane interactively since when I click on the scene, I'm rotating the geometry. It's like the arrow is "behind" the interactor which allows me to rotate the geometry.

In particular, the code that I have tried to use in my workflow, together with the code shown above, is the following.
```
def myCallback(obj, event):
    global slice_plane, slice_plane_actor
    obj.GetPlane(slice_plane)
    slice_plane_actor.VisibilityOn()

planeWidget.AddObserver("InteractionEvent", myCallback)
```

Thank you very much in advance for the help.

Greetings.

**Comment by jourdain** [comment]
You should first start testing your code with the `VtkRemoteView` to skip any limitation or corner case related to the `VtkLocalView`.

**Reply by loisvazquezind** [reply]
Hello. I am sorry for leaving this thread unanswered for so long. The widget works fine with VtkRemoveView. I even tried the DistanceWidget and the same thing happened. Why this occurs? As far as I know, the VtkLocalView renders everything on the client machine and the VtkRemoteView is constantly mapping between the client machine and the server. Are there widget mapping limitations between VTK/C++ and vtk.js?

At this moment, I am using the following packages:
```
trame==3.3.0
trame-components==2.2.1
trame-vtk==2.6.1
trame-vuetify==2.3.1
trame-router==2.1.0
vtk==9.3.0
```

**Reply by jourdain** [reply]
Yes, widgets are not getting handled right now with the VtkLocalView. Widgets are trickier than the rest, as behavior needs to be associated with them and we don't have a 1-to-1 mapping with the C++ implementation. This will change once we create a wasm implementation for the VtkLocalView, but we are not there yet.

**Reply by loisvazquezind** [reply]
Understood! Thank you for your quick response. I will look forward to that functionality.
